README
Helios version 2.60
===================

INSTALLATION

To install this module type the following:

   export HELIOS_INI=/path/to/helios.ini
   perl Makefile.PL
   make
   make test
   make install

If you don't set the $HELIOS_INI env var, 'make test' will skip its 
configuration testing step.

See the INSTALL file in this distribution for more detailed installation 
instructions.


DEPENDENCIES

This module requires these other modules and libraries:

Config::IniFiles
DBD::mysql
DBI
Data::ObjectDriver
Error
TheSchwartz
XML::Simple


COPYRIGHT AND LICENSE

This version of the software is based on Helios 2.23, which is 
Copyright (C) 2007-8 by CEB Toolbox, Inc., and is available from 
CPAN at:  http://search.cpan.org/~lajandy/Helios-2.23/

Portions of this software are Copyright (C) 2012-3 by Logical Helion, LLC
where noted.

Portions of this software are Copyright (C) 2009-12 by Andrew Johnson
where noted.

Portions of this software are Copyright (C)2007-8 CEB Toolbox, Inc.

Portions of this software are Copyright (C) 2005-6 by Six Apart, 
cpan@sixapart.com, where noted.

Portions of this software are Copyright (C) 2007, Six Apart Ltd, 
cpan@sixapart.com, where noted.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.0 or,
at your option, any later version of Perl 5 you may have available.

See the LICENSE file included in this distribution for more details.



Changes
Revision history for Perl distribution Helios.

2.82  Fri Oct 17 22:08:18 2014
	- Helios::Config - fix for [RT98284].  The getParam() method did 
	  not take the given host into account when deciding whether to 
	  retrieve a parameter value from the database or just use the 
	  cached version.
	- helios_job_submit.pl - fix for [RT98845].  The 
	  validateParamsXML() function was not returning undef if the 
	  XML arguments did not validate.  
	- Helios::Error - Added JobTypeError and ObjectDriverError to the
	  list of loaded exceptions.
	- Helios::Error::*, Bundle::Helios - Added 5.008, strict, and 
	  warnings pragmas for Kwalitee.
	- Documentation updates to helios_job_submit.pl (thanks @jscook!),
	  helios_jobtype_info, helios_service_status
	- Refactored eg/ examples to use Helios::App::* namespace, 
	  preventing CPAN namespace pollution.

2.81  Sat Aug 16 08:17:30 2014
	- Added helios_jobtype_info command.
	- Added -H as a shortcut for --hostname for all helios_config_* 
	  commands.  The helios_service_status command already had this
	  option.
	- helios_service_status:  Fix for [RT94095].  If a service does 
	  not report a status, helios_service_status will display '(none)'
	  in the Version: field.
	- helios_config_set - helios_config_set now issues a response
	  message if the config parameter was set successfully. 
	- Helios::Logger - added get/setService accessor methods to help
	  clarify the service/job/jobtype relationship.  Even though the
	  original accessors are named get/setJobType, the information 
	  being recorded is actually the service name.  Helios::Logger 
	  classes can get jobtype information from the Helios::Job 
	  object itself (that's what Helios::Logger::Internal does).  The 
	  old get/setJobType() methods will continue to work identically,
	  but will be deprecated in the next major release.
	- Helios::Config - updated POD to remove all mention of the old 
	  get/setServiceName() accessor pair to reduce confusion.
	- Helios::Job - added get/setPriority accessor methods, which were 
	  mistakenly left out of release 2.80.  

2.80  Sat Mar 15 21:19:36 2014

	[New Features]
	- New Jobtype System:  Before Helios 2.80, the jobtype of a job 
	  had to be the same name as the service class that ran the job. 
	  If you had a service class you wanted to run multiple jobtypes,
	  name a jobtype something different than the service class, or 
	  divide a workload in a complex way, you had to create multiple 
	  service classes by creating empty subclasses of your service 
	  class. This could lead to a lot of boilerplate code to generate
	  and manage. With the new Helios::JobType class and 
	  helios_jobtype_add command, new jobtypes can be created as 
	  necessary for your jobs, and the new --jobtypes option of 
	  helios.pl allows your service to work with whatever jobtype(s) 
	  you need, regardless of the service's name. This also eliminates
	  the need for the HELIOS_CLASS_MAP table, a virtually 
	  undocumented feature only used by the submitJob.pl CGI. This 
	  separation of services and jobtypes will also allow more advanced
	  job queuing features in the future. 
	- Worker Launch Patterns:  Previous versions of Helios launched 
	  workers to service jobs in a one-at-time manner.  With Helios 
	  2.80, the algorithm used to determine the number of workers to
	  launch is configurable, with 3 algorithms to choose from: 
	  "linear", "dynamic", and "optimistic". "Linear" is the old, 
	  one-at-time algorithm. The "dynamic" pattern attempts to keep 
	  as many workers running as there are available jobs in the queue,
	  adjusting the number of workers launched based on available job 
	  count. The "optimistic" pattern launches as many workers as there
	  are available jobs in the queue, regardless of the number of 
	  workers already running (limited by the MAX_WORKERS limit). The
	  default is "linear" for backward compatibility reasons. Most 
	  Helios services should work well with the "dynamic" pattern, and
	  it has been extensively tested in production environments. The 
	  "optimistic" pattern, while theoretically spooling up workers the
	  fastest, is best used with services with large numbers of jobs of
	  short duration (1-2 sec or less), as shared resources could be 
	  taxed and job contention may result. See the 
	  Helios::Configuration POD for more information.
	- New Commands:  New shell commands greatly reduce the need to 
	  consult the Helios::Panoptes web interface or touch the Helios
	  database directly with SQL commands: 
	  helios_config_get/set/unset 
	  helios_job_status 
	  helios_job_info 
	  helios_service_status 
	  Additionally, the new helios_config_import command allows you to 
	  bulk-import configuration parameters from an INI-style config 
	  file, easing deployment of services with lots of config params. 
	- Helios::TS:  Helios no longer relies directly on TheSchwartz job
	  queuing system, but on its own Helios::TS superset system.  This
	  allows new Helios-specific features to be added (including some 
	  of the changes listed above).
	- Helios::Service->JobLockInterval():  This new method allows your
	  service to specify the amount of time jobs are locked in the job 
	  queue. This was previously set at 1 hour unless you reached into
	  the TheSchwartz layer and redefined the grab_for() setting, which
	  caused problems if you wanted to use MaxRetries() and 
	  RetryInterval() to retry failed problems more frequently than 
	  once an hour.  If you wish to retry failed jobs more frequently 
	  than once an hour, define a JobLockInterval() method in your 
	  service class with the same value as your RetryInterval(). 
	  ALWAYS REMEMBER to make sure these intervals always allow enough
	  time to complete or fail a job; otherwise, two different workers
	  can end up running the same job simultaneously. 
	- Helios::Config:  New get/set/unsetParam() methods retrieve 

cgi-bin/submitJob.pl
#!/usr/bin/env perl

use 5.008;
use strict;
use warnings;
use CGI qw(:cgi -debug);
use CGI::Carp 'fatalsToBrowser';
$CGI::POST_MAX=1024 * 1000;		# 1MB maximum request size 
$CGI::DISABLE_UPLOADS = 1;		# the mechanisms we're using aren't the upload
								# the file upload mechanism
use Helios::Job;
use Helios::Service;
use Helios::LogEntry::Levels ':all';

our $VERSION = '2.41';

# grab the CGI params (we may not have any)
my $cgi = CGI->new();
my $type = $cgi->param('type');
my $params = $cgi->param('params');

# setup Helios service for config/logging/etc.
my $Service = new Helios::Service;
$Service->prep() or die($Service->errstr);
my $HeliosConfig = $Service->getConfig();

# we're going to fill these in
my $jobClass;
my $argXML;
my $job;
my $jobid;

# we were either given a bunch of XML in POSTDATA
# or a multipart MIME-encoded form
if ( defined($type) && defined($params) ) {
	# ok, the submission was made old-school,
	# via multi-part form
	$jobClass = lookupJobClass($type);
	$argXML = $params;
} elsif ( defined($cgi->param('POSTDATA')) ) {
	# submission was just a POSTed XML stream
	# parse the XML to get the type, then
	# use the type to get the job class
	$argXML = $cgi->param('POSTDATA');
	$type = parseArgXMLForType($argXML);
	$jobClass = lookupJobClass($type);
} else {
	$Service->logMsg(LOG_ERR, 'submitJob.pl ERROR: submit request failed to submit job type and/or arguments');
	die('submitJob.pl ERROR: submit request failed to submit job type and/or arguments');
}

# we have job argument XML and the class it's destined for
# so it's submission time!
eval {
	$job = Helios::Job->new();
	$job->setConfig($HeliosConfig);
	$job->setFuncname($jobClass);	
	$job->setArgXML($argXML);
	$jobid = $job->submit();
} or do {
	my $E = $@;
	$Service->logMsg(LOG_ERR, $E);
	die($E);
};

# if we encountered no errors, the job was submitted successfully
# send a nice response message to the client
print $cgi->header('text/xml');
print <<RESPONSE;
<?xml version="1.0" encoding="UTF-8"?>
<response>
<status>0</status>
<jobid>$jobid</jobid>
</response>
RESPONSE

# DONE!



sub lookupJobClass {
	my $type = shift;
	my $dbh = $Service->dbConnect();
	my $r = $dbh->selectrow_arrayref("select job_class from helios_class_map where job_type = ?", 
			undef, $type);
	if ( defined($r) && defined($r->[0]) ) {
		return $r->[0];
	} else {
		return $type;
	}
}


sub parseArgXMLForType {
	my $parsedXml = Helios::Job->parseArgXML($_[0]);
	my $type;
	if ( defined($parsedXml->{job}->[0]->{type}) ) {
		return $parsedXml->{job}->[0]->{type};
	} else {
		$Service->logMsg(LOG_ERR,"submitJob.pl ERROR: type attribute not specified in XML stream");
		die("submitJob.pl ERROR: type attribute not specified in XML stream");
	}
}


=head1 NAME

submitJob.pl - CGI script to receive jobs for Helios via HTTP POST

=head1 DESCRIPTION

Besides the built-in Perl job submission API and the helios_job_submit.pl 
command line utility, job can be submitted to Helios via an HTTP POST 
request by using the submitJob.pl CGI program.  

=head1 SUPPORTED FORMATS

The job submission request can be submitted either in a form-encoded request or
a simple stream of XML POSTDATA.

=head2 FORM ENCODED

You can submit a job through a form-encoded request.  There are two 
form fields:

=over 4

=item type

The job type, ie the service class that will perform the job

=item params

The job arguments, ie the actual parameters of the job, in XML format

=back

You can submit requests in this format via a browser using the following 
HTML form as a guide:

 <form method="post" action="http://host/cgi-bin/jobSubmit.pl">
 <b>Job Type (Service Class)</b> <br />
 <input type="text" name="type" />
 <br />
 <br />
 <b>Job Arguments</b> <br />
 <textarea rows="10" cols="80" name="params"></textarea>
 <br />
 <input type="submit" value="Submit Job" />
 </form>

The params field should contain the job arguments in XML format, like the 
following example:

 <job>
 	<params>
 		<arg1>value1</arg1>
 		<id>value2</id>
 		<email>root@somewhere.com</email>
 	</params>
 </job>

The <arg1>, <id>, etc. tags will be parsed into Perl hash values by Helios 
when the job is run.  Inside the <params> section, the tag names are up to you; 
though Helios does checks to make sure the job XML is well-formed, there is no 
DTD that the text is validated against.  Thus, you have a certain amount of 
freedom when deciding what information to pass to Helios as a job argument.  
However, it is generally a good idea to keep things short and simple.  

=head2 XML POSTDATA

Instead of encoding your request as a form, you can simply POST a stream of 
XML text.  In this type of request, the job argument XML is in the same format 
as above, with the addition that the job type is specified as a property of the 
<job> tag:

 <?xml version="1.0" encoding="UTF-8"?>
 <job type="Helios::TestService">
 	<params>
 		<arg1>a job submitted via HTTP</arg1>
 		<note>this one was posted as an XML stream</note>
 	</params>
 </job>

To POST the above XML stream with Perl using LWP::UserAgent, use the following 
script as an example:

 #!/usr/bin/env perl 
 use strict;
 use warnings;
 use LWP::UserAgent;
 use HTTP::Request::Common;
 my $ua = LWP::UserAgent->new();
 my $jobXML = <<ENDXML;
 <?xml version="1.0" encoding="UTF-8"?>
 <job type="Helios::TestService">
 	<params> 
 		<arg1>a job submitted via HTTP</arg1>
 		<note>this one was posted as an XML stream</note>
 	</params>
 </job>
 ENDXML

 my $r = $ua->request(POST 'http://localhost/cgi-bin/submitJob.pl',
 		Content_Type => 'text/xml',
 		Content => $jobXML
 );
 print $r->as_string;

=head1 RESPONSE

Regardless of which encoding method you use for your request, if your job was 
submitted successfully you'll receive a response like the following:

 HTTP/1.1 200 OK
 Connection: close
 Date: Fri, 16 Dec 2011 04:24:30 GMT
 Server: Apache/2.2.14 (Ubuntu)
 Vary: Accept-Encoding
 Content-Type: text/xml; charset=ISO-8859-1
 Client-Date: Fri, 16 Dec 2011 04:24:30 GMT
 Client-Peer: 127.0.0.1:80
 Client-Response-Num: 1
 Client-Transfer-Encoding: chunked

 <?xml version="1.0" encoding="UTF-8"?>
 <response>
 <status>0</status>
 <jobid>42</jobid>
 </response>

If your job was submitted successfully, the <status> section will contain a 0,
and the <jobid> section will contain the jobid of the submitted job.  If there 
was an error during the job submission process, a 500 Internal Server Error 
will be returned.

=head1 FUNCTIONS

=head2 lookupJobClass($type)

Given a job type, lookupJobClass() queries the HELIOS_CLASS_MAP table to 
determine if a Helios service class is associated with that job type.  If 
there is, the function returns that service class.  If it isn't, the original 
type is assumed to be the name of the desired service class, and it is 
returned to the caller instead.

=head2 parseArgXMLForType($xmlstring)

Given a string of job arguments in XML format, parseArgXMLForType() 
returns the value of the type attribute of the <job> tag, if specified.

=head1 SEE ALSO

L<helios_job_submit.pl>, L<helios.pl>, L<Helios::Service>, L<Helios::Job>

=head1 AUTHOR

Andrew Johnson, E<lt>lajandy at cpan dot orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011-2 Andrew Johnson.

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.0 or,
at your option, any later version of Perl 5 you may have available.

=head1 WARRANTY

This software comes with no warranty of any kind.

=cut

